File: fiboakin.py
#https://www.codewars.com/kata/5772382d509c65de7e000982/train/python

u_dict = {
    1:1,
    2:1,
        }

def u(n):
    if n in u_dict:
        return u_dict[n]
    else:
        t = u(n-u(n-1)) + u(n-u(n-2))
        u_dict[n] = t
        return t

def length_sup_u_k(n, k):
    count = 0
    for i in range(1,n+1):
        if u(i) >= k:
            count += 1
    return count

def comp(n):
    count = 0
    for i in range(2,n+1):
        if u(i) < u(i-1) :
            count += 1
    return count
    

print(length_sup_u_k(3332,973))
print(comp(74626))

File: fiboakin.rs
// https://www.codewars.com/kata/5772382d509c65de7e000982/train/rust

use std::collections::HashMap;

fn u(n:i32, u_dict: &mut HashMap<i32,i32>) -> i32 {
    if n == 1 {
        return 1;
    } else if n == 2 {
        return 1;
    } else if u_dict.contains_key(&n) {
        return *(u_dict.get(&n).unwrap());
    } else {
        let t = u(n-u(n-1,u_dict), u_dict) + u(n-u(n-2,u_dict),u_dict);
        u_dict.insert(n,t);
        return t;
        }
}

fn length_sup_uk(n: i32, k: i32) -> i32 {
    let mut count = 0;
    let mut u_dict = HashMap::new();
    for i in 1..=n {
        if u(i, &mut u_dict) >= k {
            count += 1;
        }
    }
    count
}

fn comp(n: i32) -> i32 {
    let mut count = 0;
    let mut u_dict = HashMap::new();
    for i in 2..=n {
        if u(i,&mut u_dict) < u(i-1, &mut u_dict) {
            count += 1;
        }
    }
    count
}